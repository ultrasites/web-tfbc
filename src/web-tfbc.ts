import * as fs from "fs";
import { WebTFBCConfig, WebTFBCDefaultConfig } from "./web-tfbc.types";

export class WebTFBC {
  config: WebTFBCConfig;

  constructor(config: WebTFBCConfig = WebTFBCDefaultConfig) {
    this.config = config;
  }

  private _isExpired(fileBirthday: number): boolean {
    const currentTimestamp = new Date().valueOf();
    const fsTimestamp = new Date(fileBirthday).valueOf();
    return currentTimestamp - fsTimestamp > this.config.expired;
  }

  private _clear(key: string) {
    fs.unlinkSync(this.config.filePath + key);
  }

  private _get<T>(key: string): T | boolean | void {
    try {
      const stats = fs.statSync(this.config.filePath + key);

      if (!stats.isFile()) {
        return false;
      }

      if (this._isExpired(stats.birthtimeMs)) {
        this._clear(key);
        return false;
      } else {
        try {
          const content = fs.readFileSync(this.config.filePath + key, {
            encoding: this.config.encoding,
          });
          return content as unknown as T;
        } catch (error) {
          console.log(error);
        }
      }
    } catch (_e) {
      return false;
    }
  }

  async cache<T>(key: string, content: T) {
    try {
      await fs.promises.access(this.config.filePath);
    } catch {
      fs.mkdirSync(this.config.filePath, { recursive: true });
    }
    const cached = this._get<T>(key);
    let data = "";

    if(content instanceof HTMLElement) {
      data = content.innerHTML;
    } else {
      data = content as unknown as string;
    }

    if (cached !== false) {
      return cached as T;
    } else {
      fs.writeFileSync(
        this.config.filePath + key,
        data,
        { encoding: this.config.encoding }
      );
      return content;
    }
  }
}
