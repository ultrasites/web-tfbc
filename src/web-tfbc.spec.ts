import { readFileSync, unlink } from "fs";
import { WebTFBC } from "./web-tfbc";

describe("web-tfbc", () => {
  it("should cache a simple string", async () => {
    const cache = new WebTFBC();
    const content = await cache.cache<string>("test", "test");
    expect(content).toEqual("test");
  });

  it("should cache simple string as expired", async () => {
    const cache = new WebTFBC({
      filePath: "./src/testfiles/",
      encoding: "utf-8",
      expired: 1000,
    });
    const content = await cache.cache<string>("expiredFile", "expiredFile");

    expect(content).toEqual("expiredFile");
  });

  it("should cache html file", async () => {
    const htmlFile = readFileSync("./src/testfiles/test.html", {
      encoding: "utf-8",
    });

    const cache = new WebTFBC();
    const content = await cache.cache<string>("htmlTest", htmlFile);

    expect(content).toMatchSnapshot();
  });

  it("should cache dom", async () => {
    document.body.innerHTML =
      '<div id="cache">' +
      '  <span id="username" />' +
      '  <button id="button" />' +
      "</div>";

    const cache = new WebTFBC();
    const content = await cache.cache<HTMLElement>(
      "htmlElement",
      document.getElementById("cache")!
    );

    expect(content).toMatchSnapshot();
  });

  afterEach(async () => {
    await unlink("./web-tfbc-cache", () => {});
  });
});
