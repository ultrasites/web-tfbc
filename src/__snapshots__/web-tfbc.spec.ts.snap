// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web-tfbc should cache dom 1`] = `"  <span id=\\"username\\">  <button id=\\"button\\"></button></span>"`;

exports[`web-tfbc should cache html file 1`] = `
"<html>
  <head></head>

  <body>
    Cached!
  </body>
</html>
"
`;
